### PR TITLE
[SPARK-6325] [core,yarn] Do not change target executor count when killing executors.

### PR DESCRIPTION
**Pull request by vanzin**
The dynamic execution code has two ways to reduce the number of executors: one
where it reduces the total number of executors it wants, by asking for an absolute
number of executors that is lower than the previous one. The second is by
explicitly killing idle executors.

YarnAllocator was mixing those up and lowering the target number of executors
when a kill was issued. Instead, trust the frontend knows what it's doing, and kill
executors without messing with other accounting. That means that if the frontend
kills an executor without lowering the target, it will get a new executor shortly.

The one situation where both actions (lower the target and kill executor) need to
happen together is when user code explicitly calls `SparkContext.killExecutors`.
In that case, issue two calls to the backend to achieve the goal.

I also did some minor cleanup in related code:
- avoid sending a request for executors when target is unchanged, to avoid log
  spam in the AM
- avoid printing misleading log messages in the AM when there are no requests
  to cancel
- fix a slow memory leak plus misleading error message on the driver caused by
  failing to completely unregister the executor.

### CODE REVIEW

**Comment by shaneknapp** [comment]
jenkins, test this please

**Comment by vanzin** [comment]
Talked to @sryza offline and we decided to go with a simpler fix; after some thinking, the only bad effect of this approach is that if the frontend (= driver) kills executors without lowering the target count, it will get new executors allocated. But we're pretty sure the code doesn't do that.

**Comment by vanzin** [comment]
Jenkins, test this please.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 1]
```diff
@@ -371,6 +376,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val actorSyste
         logWarning(s"Executor to kill $id does not exist!")
       }
     }
+    // Killing executors means effectively that we want less executors than before, so also update
+    // the target number of executors to avoid having the backend allocate new ones.
+    val newTotal = numExistingExecutors + numPendingExecutors - executorsPendingToRemove.size
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: I think you need parentheses to make the part on the next line included in the statement

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 2]
```diff
@@ -211,6 +211,7 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val actorSyste
           // This must be synchronized because variables mutated
           // in this block are read when requesting executors
           CoarseGrainedSchedulerBackend.this.synchronized {
+            addressToExecutorId -= executorInfo.executorAddress
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: Is this a separate bug that you noticed?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 3]
```diff
@@ -340,7 +341,11 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val actorSyste
     }
     numPendingExecutors =
       math.max(numExecutors - numExistingExecutors + executorsPendingToRemove.size, 0)
-    doRequestTotalExecutors(numExecutors)
+    if (numPendingExecutors > 0) {
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: What's the issue with requesting here?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 2 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: Yes. It was in the previous summary but I overwrite it in my edit. Will restore the explanation. (Comment also valid for other things you asked about.)

**Comment by srowen** [comment]
I don't know a lot about this code, but based on the description and fix, it looks reasonable. Since Sandy created the bit of code being modified, I trust his review on this one. It looks like @sryza has a few outstanding questions here. Once those are satisfied this is looking good. 

Any other thoughts from YARN folk?

**Comment by vanzin** [comment]
@sryza's questions should be answered in the updated PR summary.

**Comment by sryza** [comment]
Planning to look at this later today

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 3 above)
Comment (ksakellis) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: I had the same question. Why is this change necessary?

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: Let me point you guys again at the PR summary.

**Comment by ksakellis** [comment]
I'm not very familiar with the code but it looks like a pretty sane change from what I can tell.

**Comment by sryza** [comment]
If I understand correctly, this has the same drawback as the earlier patch: the driver can't communicate to the AM that it wants fewer executors than are currently running.  This means that if executors die naturally, the AM will request new ones unnecessarily.

We had discussed an approach where killing executors basically doesn't affect any of the bookkeeping for pending / target num executors.  Were there other drawbacks with that?

**Comment by vanzin** [comment]
> the driver can't communicate to the AM that it wants fewer executors than are currently running

It can. But the AM won't kill running executors if the target count goes below the number of running executors. It will just set the target to the new, lower value, so that when the driver kills executors explicitly, the AM won't start new ones.

> killing executors basically doesn't affect any of the bookkeeping for pending / target num executors

That's exactly what this patch does - on the AM side. What we had discussed is that on the driver side, you do need to update the target number when you kill an executor explicitly by calling `SparkContext.killExecutors`. Otherwise the AM will just start new ones for you.

**Comment by sryza** [comment]
To be more specific, I'm wondering whether, in `ExecutorAllocationManager`, we should be removing `executorsPendingToRemove.size` from the `targetNumExecutors()` calculation.

**Comment by vanzin** [comment]
That method looks correct given the scaladoc describing it.

Note that user code has two ways of affecting that method: `SparkContext.requestExecutors`, which bumps `numExecutorsPending`, and `SparkContext.killExecutors`, which updates `executorsPendingToRemove`. So the implementation looks correct from `SparkContext`'s point of view.

**Comment by vanzin** [comment]
(Just for completeness: `SparkContext` actually doesn't directly affect the bookkeeping in `ExecutorAllocationManager`, which can be seen as a separate issue. Meaning my last comment is not entirely accurate, but the current bookkeeping in `ExecutorAllocationManager` seems correct.)

**Comment by sryza** [comment]
I have two main concerns about this patch.

The first is that I think the logic in `CoarseGrainedSchedulerBackend` and `ExecutorAllocationManager` is overcomplicated.  I think we might be able to remove `executorsPendingRemoval.size` from the `targetNumExecutors` in both of these places.  I also think it's confusing that `pendingNumExecutors` can get out of sync (it can be negative in `ExecutorAllocationManager`, but not in `CoarseGrainedSchedulerBackend`).  However, this patch isn't really the source of additional complexity. 

My second is that we now make two RPCs to the AM when we want to kill a container.  Ideally we would have two paths: one that maintains the existing behavior of `killExecutors` for users and includes the `requestTotalExecutors` RPC.  And another meant for consumption by `ExecutorAllocationManager` that calls `killExecutor` directly.

That said, I've spent a while today trying to think of situations where this patch would lead to incorrect behavior and have become convinced that it should work.  As the patch fixes a critical bug and doesn't introduce additional complexity, I'd be OK with merging it and fixing the other issues later.

**Comment by vanzin** [comment]
+1 to this code being confusing and overly complicated. There are 3 places tracking executor state (ExecutorAllocationManager, CoarseGrainedSchedulerBackend and YarnAllocator), not to mention all the other places where the RPCs are plumbed through. It's a maze just to find out what calls out to what, and which is the right state at any moment.

But the patch is targeted at fixing the bug and being as non-intrusive as possible.

**Comment by srowen** [comment]
I will let it sit another day, but sounds like there is consensus that this does no harm and fixes a bug. Does this need to go back in to `branch-1.3`?

**Comment by vanzin** [comment]
Yes we should backport for 1.3.1 (I believe I set the bug accordingly?).

**Comment by andrewor14** [comment]
@vanzin @sryza Thanks for working on the fix. I was away for the past week and did not have the chance to review this before it went in. Regarding the code being overly complicated, the reason why the bookkeeping is done in each of the three places you pointed out is the following:
- We need to do it in `ExecutorAllocationManager` so we don't keep requesting beyond the configured maximum
- We need to do it in `CoarseGrainedExecutorBackend` because the user can bypass the dynamic scaling logic and explicitly request executors through `sc.requestTotalExecutors`.
- We need to do it in `YarnAllocator` to ensure we don't over-allocate containers, regardless of whether `sc.requestTotalExecutors` or the dynamic scaling logic is used.

My intent is also to simplify the code so as to minimize the possibility of this feature breaking again in a future release. If either of you have concrete suggestions on refactoring it for this purpose, I would love to hear them.

**Comment by vanzin** [comment]
Hi @andrewor14 ,

We don't really have concrete plans to change this in the very short term; if/when we get down to enhancing the dynamic allocation backend, cleaning up this code (so that it's easy to make sure we're not breaking anything) would probably be part of the work.

Nevertheless, it's still useful to know the motivation for having that code in there, so thanks!

**Comment by sryza** [comment]
@andrewor14 agreed that the bookkeeping needs to be done in these three places.  Didn't meant to imply that the code is deeply muddled.  To be more precise, there are two places that I think there may be opportunity for simplifying the code, one caused by this patch, one in light of this patch.
* Before this patch, the bookkeeping in CoarseGrainedSchedulerBackend wasn't used in between the ExecutorAllocationManager and the YarnAllocator.  numPendingExecutors was updated, but never read.  It would be nice to go back to basically having the ExecutorAllocationManager communicate directly with the YarnAllocator.
* In YarnAllocator, this patch was able to remove the effect of killExecutor on targetNumExecutors.  I'm wondering whether a similar change in ExecutorAllocationManager is possible, though this warrants further investigation.

Will try to look deeper when I have the time.

**Comment by prakhar-jauhari** [comment]
Hey @vanzin 
Thanks in advance.
I ran in to a issue lately : https://issues.apache.org/jira/browse/SPARK-11181, while debugging i came across this pull request. 

I am new to this side of code, i want to understand the scenario's as why do we reduce the total number of executors on a host expiry.

Regards,
Prakhar.

**Comment by vanzin** [comment]
@prakhar-jauhari asking questions in a closed PR is confusing. If you have doubts about the code, I suggest asking dev@spark.apache.org.

As for you question, the answer is, we don't.

**Comment by prakhar-jauhari** [comment]
@vanzin thanks for the response. And not posting to a closed PR (noted!!)

In my case spark driver is reducing the total number of executors on host expiry. 
I had already posted my query in dev@spark.apache.org. I would be great if u can have a look at it, as it depends on your changes in CoarseGrainedSchedulerBackend.scala in this PR.
